### DISCUSSION TITLE
Library change breaking one of several apps

### DISCUSSION THREAD

**Discussion by KarmaCop213** [nrwl/nx#11347] [General]
Long story short: on a monorepo with several apps, a change on a shared library broke one of the apps.

My question to the community is the following: how can we prevent the above?

There's unit tests on 90% of the code. Some apps have more e2e tests than others. But I don't know if we should rely that much on e2e testing to catch this kind of problems. 

If I had separated repos I could control the library version and only update it after testing using the new library version. With monorepos I don't have that kind of control.

**Comment by rfprod** [comment]
@KarmaCop213 From my point of view, maybe integration tests are missing in your case. Do you use integration tests?
Unit tests test individual units. Integration tests validate interoperation of units. End to end tests aka user acceptance tests validate a product from the user perspective.
Generally, if something like what you described happens, the problem is fixed and an integration test is added to make sure that this thing doesn't break again without failing the integration test.

**Reply by KarmaCop213** [reply]
Although I understand your point of view, the idea is preventing that from happening, not creating a test after it fails.

Creating more integration test can be a solution, but we end up having the same issues, something not being tested can fail without notice if the shared library is changed. It's virtually impossible to test everything 100%.

The whole idea here is to provide the management 100% certainty that a change on a shared library won't cause issues on the apps that consume it. I can't see testing (integration or E2E) being the reliable solution we want and need for this kind of problem.

I just wonder how other companies manage this kind of issue, do they just assume that apps can fail? And if they do, do they use monorepos for highly critical apps such as medical treatment, health monitoring, etc.
